src/tqec/computation/cube.py

```diff
@@ -0,0 +1,183 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from dataclasses import astuple, dataclass
+from enum import Enum
+
+from tqec.computation.zx_graph import ZXKind, ZXNode
+from tqec.exceptions import TQECException
+from tqec.position import Direction3D, Position3D
+
+
+class ZXBasis(Enum):
+    """Z or X basis."""
+
+    Z = "Z"
+    X = "X"
+
+    def with_zx_flipped(self) -> ZXBasis:
+        """Get the basis with the Z/X flipped."""
+        return ZXBasis.Z if self == ZXBasis.X else ZXBasis.X
+
+    def __str__(self) -> str:
+        return self.value
+
+
+class CubeKind(ABC):
+    """Base class for the cube types."""
+
+    @abstractmethod
+    def to_zx_kind(self) -> ZXKind:
+        """Convert the cube kind to a ZX node kind."""
+        pass
+
+
+@dataclass(frozen=True)
+class ZXCube(CubeKind):
+    """Cube kind representing the cube surrounded by all X/Z basis walls."""
+
+    x: ZXBasis
+    y: ZXBasis
+    z: ZXBasis
+
+    def as_tuple(self) -> tuple[ZXBasis, ZXBasis, ZXBasis]:
+        return astuple(self)
+
+    def __post_init__(self) -> None:
+        if self.x == self.y == self.z:
+            raise TQECException(
+                "The cube kind with all the same basis walls is not allowed."
+            )
+
+    def __str__(self) -> str:
+        return f"{self.x.value}{self.y.value}{self.z.value}"
+
+    @staticmethod
+    def all_kinds() -> list[ZXCube]:
+        """Return all the possible `ZXCube` kinds."""
+        return [ZXCube.from_str(s) for s in ["ZXZ", "XZZ", "ZXX", "XZX", "XXZ", "ZZX"]]
+
+    @staticmethod
+    def from_str(string: str) -> ZXCube:
+        """Create a cube kind from the string representation."""
+        return ZXCube(*map(ZXBasis, string.upper()))
+
+    @property
+    def cube_basis(self) -> ZXBasis:
+        """Return the basis of the cube.
+
+        A cube has only one Z/X basis wall is a Z/X basis cube.
+        """
+        if sum(basis == ZXBasis.Z for basis in astuple(self)) == 1:
+            return ZXBasis.Z
+        return ZXBasis.X
+
+    def to_zx_kind(self) -> ZXKind:
+        return ZXKind(self.cube_basis.value)
+
+    @property
+    def normal_direction(self) -> Direction3D:
+        """Get the direction in which the wall basis is the same as the cube
+        basis."""
+        return Direction3D(astuple(self).index(self.cube_basis))
+
+    @staticmethod
+    def from_normal_basis(basis: ZXBasis, direction: Direction3D) -> ZXCube:
+        """Create a cube kind with the given normal basis and direction."""
+        bases = [basis.with_zx_flipped() for _ in range(3)]
+        bases[direction.value] = basis
+        return ZXCube(*bases)
+
+    @property
+    def is_regular(self) -> bool:
+        """Check if the cube is a regular cube."""
+        return self.x != self.y
```

Comment:
Same remark about the use of the world "regular" that does not provide enough information to the caller in my opinion.